### PR TITLE
feat: SUP-2631: ERC1155s heavy refactor

### PR DESCRIPTION
**Pull request by shortdoom**
SuperForm's standalone ERC1155 implementation.

Problem: We have singleId approve (and range approve in future) logic which can't operate with implementation of solmate & openzeppelin's safeTransferFrom() 

Solution: We implement ERC1155 functions according to the IERC1155 standard ourselves. We host all of the necessary external functions and libraries in `src/utils` avoiding reduntant imports.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -4,7 +4,3 @@
 [submodule "lib/solmate"]
```
Comment (0xTimepunk) [inline .gitmodules:4]: Do we still need solmates library here?

File: src/ERC1155s.sol [hunk 2]
```diff
@@ -274,13 +308,121 @@ abstract contract ERC1155s is ERC1155 {
...
 
     /// @dev Used to construct return url
     function _baseURI() internal view virtual returns (string memory);
+
+    /*//////////////////////////////////////////////////////////////
+                              ERC165 LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /// @dev Implementation copied from solmate/ERC1155
+    function supportsInterface(bytes4 interfaceId) public view virtual returns (bool) {
```
Comment (0xTimepunk) [inline src/ERC1155s.sol:324]: I'm not sure if we need to check for all those ids, right?

**Review by 0xTimepunk — DISMISSED** [review]
Looking great! Just the decreaseAllowance I couldn't comment, but looks like it can directly re-use the internal _ decreaseAllowance inside of it, instead of repeating the logic?

File: src/ERC1155s.sol (hunk 2 above)
Comment (shortdoom) [reply src/ERC1155s.sol:324]: ERC1155 EIP doc specifies it:

_Smart contracts implementing the ERC-1155 standard MUST implement the ERC-165 supportsInterface function and MUST return the constant value true if 0xd9b67a26 is passed through the interfaceID argument._

**Comment by shortdoom** [comment]
> Looking great! Just the decreaseAllowance I couldn't comment, but looks like it can directly re-use the internal _ decreaseAllowance inside of it, instead of repeating the logic?

Same situation as with `safeTransferFrom()` and internal `_safeTransferFrom()`. Because we have single Id approves we can't re-use msg.sender, it needs to be first assigned at external call and then passed as an argument to internal call.

**Review by 0xTimepunk — DISMISSED** [review]

**Review by 0xTimepunk — APPROVED** [review]

**Pull request merged**